### PR TITLE
playlists and favorites

### PR DESCRIPTION
**Pull request by achimmihca**
### What does this PR do?

Implement #158.

- Added playlist parser for the format that I suggested. There are also some tests for parsing lines.
- new PlaylistManager scans the playlist folder. A special "Favorites" playlist is always created.
- SongSelectScene has a little heart icon to indicate whether a song is a favorite. Click this icon or hit `F` to toggle the favorite state.
- A button in the lower right corner with a label shows the current playlist and is used to iterate playlists. You can also hit `Shift+F` to switch to / from the special favorite playlist.
- The song selection shows only songs of the selected playlist. Thereby, a special "All songs" playlist is selected as default.

I wanted the playlist format to be easily written with normal text editors. This is why I tried to preserve all lines when reading and writing playlists (e.g. comments). Therefor, I modeled the playlist class as a list of lines, some of which are song entries.

### Closes Issue(s)

#158

### CODE REVIEW

File: UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                    && (targetToken == Token.Artist || targetToken == Token.Separator))
+                {
+                    targetToken = Token.Title;
+                }
+                else if (c != '\\' || (c == '\\' && lastChar == '\\'))
+                {
+                    switch (targetToken)
+                    {
+                        case Token.Artist:
+                            artist += c;
```
Comment (basisbit) [inline UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs:91]: Codacy Issue found: [Use a StringBuilder instead.](https://app.codacy.com/manual/UltraStar-Deluxe/Play/pullRequest?prid=5574890)

File: UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                            break;
+                        case Token.Title:
+                            title += c;
```
Comment (basisbit) [inline UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs:94]: Codacy Issue found: [Use a StringBuilder instead.](https://app.codacy.com/manual/UltraStar-Deluxe/Play/pullRequest?prid=5574890)

File: UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                    }
+                }
+                else if (!insideQuote && c == separator
+                    && (targetToken == Token.Artist || targetToken == Token.Separator))
+                {
+                    targetToken = Token.Title;
+                }
+                else if (c != '\\' || (c == '\\' && lastChar == '\\'))
+                {
+                    switch (targetToken)
```
Comment (basisbit) [inline UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs:88]: Codacy Issue found: [Add a 'default' clause to this 'switch' statement.](https://app.codacy.com/manual/UltraStar-Deluxe/Play/pullRequest?prid=5574890)

File: UltraStar Play/Assets/Common/Playlist/UltraStarPlaylist.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+public class UltraStarPlaylist
+{
+    private readonly List<UltraStartPlaylistLineEntry> lineEntries = new List<UltraStartPlaylistLineEntry>();
+    public IReadOnlyList<UltraStartPlaylistLineEntry> LineEntries => lineEntries;
+
+    private readonly List<UltraStartPlaylistSongEntry> songEntries = new List<UltraStartPlaylistSongEntry>();
+    public IReadOnlyList<UltraStartPlaylistSongEntry> SongEntries => songEntries;
+
+    public UltraStarPlaylist()
```
Comment (basisbit) [inline UltraStar Play/Assets/Common/Playlist/UltraStarPlaylist.cs:16]: Codacy Issue found: [Remove this redundant constructor.](https://app.codacy.com/manual/UltraStar-Deluxe/Play/pullRequest?prid=5574890)

File: UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs (hunk 3 above)
Comment (basisbit) [inline UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs:88]: Codacy Issue found: [Replace this 'switch' statement with 'if' statements to increase readability.](https://app.codacy.com/manual/UltraStar-Deluxe/Play/pullRequest?prid=5574890)

File: UltraStar Play/Assets/Common/Playlist/PlaylistManager.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        foreach (string filePath in playlistFilePaths)
+        {
+            UltraStarPlaylist playlist = UltraStarPlaylistParser.ParseFile(filePath);
+            playlists.Add(playlist);
+            playlistToFilePathMap.Add(playlist, filePath);
+
+            // This is the special playlist for the favorite songs.
+            if (Path.GetFullPath(favoritesPlaylistFile) == Path.GetFullPath(filePath))
+            {
+                favoritesPlaylist = playlist;
```
Comment (basisbit) [inline UltraStar Play/Assets/Common/Playlist/PlaylistManager.cs:95]: Codacy Issue found: [Make the enclosing instance method 'static' or remove this set on the 'static' field.](https://app.codacy.com/manual/UltraStar-Deluxe/Play/pullRequest?prid=5574890)

File: UltraStar Play/Assets/Common/Playlist/PlaylistManager.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            return favoritesPlaylistFile;
+        }
+        return playlistToFilePathMap[playlist];
+    }
+
+    private void ScanPlaylistFolder()
+    {
+        playlists = new List<UltraStarPlaylist>();
+        playlistToFilePathMap = new Dictionary<UltraStarPlaylist, string>();
```
Comment (basisbit) [inline UltraStar Play/Assets/Common/Playlist/PlaylistManager.cs:82]: Codacy Issue found: [Make the enclosing instance method 'static' or remove this set on the 'static' field.](https://app.codacy.com/manual/UltraStar-Deluxe/Play/pullRequest?prid=5574890)

File: UltraStar Play/Assets/Common/Playlist/PlaylistManager.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        if (playlist == FavoritesPlaylist)
+        {
+            return favoritesPlaylistFile;
+        }
+        return playlistToFilePathMap[playlist];
+    }
+
+    private void ScanPlaylistFolder()
+    {
+        playlists = new List<UltraStarPlaylist>();
```
Comment (basisbit) [inline UltraStar Play/Assets/Common/Playlist/PlaylistManager.cs:81]: Codacy Issue found: [Make the enclosing instance method 'static' or remove this set on the 'static' field.](https://app.codacy.com/manual/UltraStar-Deluxe/Play/pullRequest?prid=5574890)

File: UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            char lastChar = '0';
+            foreach (char c in line)
+            {
+                if (c == '"' && lastChar != '\\')
+                {
+                    insideQuote = !insideQuote;
+                    if (insideQuote)
+                    {
+                        // What has been read before for this token is now obsolete
+                        if (targetToken == Token.Artist)
```
Comment (basisbit) [inline UltraStar Play/Assets/Common/Playlist/UltraStarPlaylistParser.cs:61]: Codacy Issue found: [Refactor this code to not nest more than 3 control flow statements.](https://app.codacy.com/manual/UltraStar-Deluxe/Play/pullRequest?prid=5574890)

**Comment by achimmihca** [comment]
The PR should be fine now.

I vote to ignore / remove the rule "Make the enclosing instance method 'static' or remove this set on the 'static' field."
Using static fields is done on purpose and so far this rule was not benefitial (also in my other PR).

**Pull request merged**